r_code/estimator_drivers.R

```diff
@@ -0,0 +1,427 @@
+  # Create a list of placebos and add random noise.
+  # Note that matching from treated unit to placebo treated unit is done just
+  # once, so the 50 draws are for the same set of placebo treated units.
+  list_placebos <- lapply(seq_len(num_placebos), function(x) placebo_data)
+  noised_placebos <- furrr::future_map(
+    .x = list_placebos,
+    .f = function(z) {
+      z %>%
+        dplyr::mutate(
+          !!as.name(outcome_var) :=
+            !!as.name(outcome_var) +
+              stats::rnorm(
+                dplyr::n(), 0,
+                sd_placebo
+              ),
+          "counter_factual" =
+            !!as.name(outcome_var)
+        )
+    }
+  )
+  # Estimate the outcome series for each placebo in the list.
+  # Map the desired (default) estimators onto the placebo set.
+  placebo_estimates <- furrr::future_map(
+    .x = noised_placebos, .f = ~ .MethodEstimator(
+      data_name = .x,
+      method_name = method_name, id_var = id_var, time_var = time_var,
+      outcome_var = outcome_var, treat_indicator = treat_indicator,
+      counterfac_var = counterfac_var
+    )
+  )
+  # Estimate the ATT over the placebo sets.
+  placebo_att <- furrr::future_map(
+    .x = placebo_estimates,
+    .f = ~ ComputeTreatmentEffect(
+      pred_series_tib = .x,
+      effect_type = effect_type,
+      post_treat_only = post_treat_only,
+      pred_var = "point.pred"
+    )
+  )
+  return(placebo_att)
+}
+
+#' Compares desire method performance (RMSE) on placebo data and computes
+#'    estimates for the winner.
+#'
+#' @param method_names Vector of string names for desired methods to estimate.
+#' @param horizon Range of periods over which to display the output. Should be
+#'    non-negative here because several methods perfectly fit pre-treatment
+#'    periods, which would result in more desirable RMSE properties.
+#' @inheritParams DebiasedATT
+#'
+#' @return A long form tibble with the winning method (given input) estimates of
+#'    the outcome variable (missing outcome imputation) as well as the point
+#'    and percent effects, and a message informing the user of the winner.
+# TODO(alexdkellogg): Could create a similar function that instead returns
+# a table with Bias, RMSE, coverage of each method on placebo draws?
+AutoEstimator <- function(raw_data, method_names,
+                          num_placebos = 50, sd_placebo = 0.2,
+                          horizon = c(0, 4), effect_type = "mean",
+                          id_var = "entry",
+                          time_var = "period",
+                          treat_indicator = "treatperiod_0",
+                          outcome_var = "target",
+                          counterfac_var = "counter_factual") {
+  # Ensure the method names are entered appropriately.
+  stopifnot(method_names %in%
+    c("CausalImpact", "Gsynth", "SDID", "SDID_Uncon", "SCM"))
+  # Make sure horizon is positive. Some methods have 0 training error is why.
```

Comment:
Is the second sentence proper English?

Reply:
'Tis not.

Reply:
Though technically the third "sentence" haha.